src/internal/operators/filter.ts

```diff
@@ -5,8 +5,9 @@ import { OperatorSubscriber } from './OperatorSubscriber';
 
 /* tslint:disable:max-line-length */
 export function filter<T, S extends T>(predicate: (value: T, index: number) => value is S, thisArg?: any): OperatorFunction<T, S>;
-// NOTE(benlesh): T|null|undefined solves the issue discussed here: https://github.com/ReactiveX/rxjs/issues/4959#issuecomment-520629091
-export function filter<T>(predicate: BooleanConstructor): OperatorFunction<T | null | undefined, NonNullable<T>>;
+export function filter<T>(
+  predicate: BooleanConstructor
+): OperatorFunction<T, T extends null | undefined | false | 0 | -0 | 0n | '' ? never : T>;
```

Comment:
I feel like this is something that must exist elsewhere. `first()` for example. We probably need a type like:

```ts
type Truthy<T>  = T extends null | undefined | false | 0 | -0 | 0n | '' ? never : T
```

Reply:
What about?

```ts
type Falsy = null | undefined | false | 0 | -0 | 0n | "";

type NonFalsy<T> = T extends Falsy ? never : T;
type NonTruthy<T> = T extends Falsy ? T : never;
```

Would that be helpful anywhere else? I wouldn't mind working on that PR.